File: leetcode/algorithms/219_contains_duplicate_2/main.dart
class ContainsDuplicateII {
  /**
   * Hash Map
   *
   * Complexities:
   *   N - Size of `nums`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(N)
   */
  bool containsNearbyDuplicate(List<int> nums, int k) {
    Map<int, int> numsMap = Map();

    for (int i = 0; i < nums.length; i++) {
      if (numsMap[nums[i]] == null) {
        numsMap[nums[i]] = i;
      } else {
        if ((numsMap[nums[i]]! - i).abs() <= k) {
          return true;
        }
        numsMap[nums[i]] = i;
      }
    }

    return false;
  }


  // Solution
  /**
   * Solution 1
   *
   * Sliding Window + Hash Set
   *
   * Complexities:
   *   N - Size of `nums`
   *   K - `k`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(min(N, K))
   */
  bool solution1(List<int> nums, int k) {
    final Set<int> window = {};

    for (int i = 0; i < nums.length; i++) {
      if (i > k) {
        window.remove(nums[i - k - 1]);
      }

      if (!window.add(nums[i])) {
        return true;
      }
    }

    return false;
  }

  /**
   * Solution 2
   *
   * Hash Map
   *
   * Complexities:
   *   N - Size of `nums`
   *   - Time Complexity: O(N)
   *   - Space Complexity: O(N)
   */
  bool solution2(List<int> nums, int k) {
    final Map<int, int> lastSeen = {};

    for (int i = 0; i < nums.length; i++) {
      int num = nums[i];

      if (lastSeen.containsKey(num) && (i - lastSeen[num]!) <= k) {
        return true;
      }

      lastSeen[num] = i;
    }

    return false;
  }
}

File: leetcode/algorithms/219_contains_duplicate_2/main.py
from typing import List


class ContainsDuplicateII:
    """
    # Hash Map
    #
    # Complexities:
    #   N - Size of `nums`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(N)
    """
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        numsMap = {}

        for i in range(len(nums)):
            if numsMap.get(nums[i]) != None:
                if abs(i - numsMap[nums[i]]) <= k:
                    return True
            numsMap.update({nums[i]: i})

        return False


    # Solution
    """
    # Solution 1
    #
    # Sliding Window + Hash Set
    #
    # Complexities:
    #   N - Size of `nums`
    #   K - `k`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(min(N, K))
    """
    def solution1(self, nums: List[int], k: int) -> bool:
        seen = set()

        for i, num in enumerate(nums):
            if num in seen:
                return True

            seen.add(num)

            if len(seen) > k:
                seen.remove(nums[i - k])

        return False

    """
    # Solution 2
    #
    # Hash Map
    #
    # Complexities:
    #   N - Size of `nums`
    #   - Time Complexity: O(N)
    #   - Space Complexity: O(N)
    """
    def solution2(self, nums: List[int], k: int) -> bool:
        num_to_index = {}

        for i, num in enumerate(nums):
            if num in num_to_index and i - num_to_index[num] <= k:
                return True

            num_to_index[num] = i

        return False
